src/backend/distributed/connection/remote_commands.c

```diff
@@ -958,6 +959,10 @@ WaitForAllConnections(List *connectionList, bool raiseInterrupts)
 			waitEventSet = NULL;
 		}
 
+		pfree(allConnections);
+		pfree(events);
+		pfree(connectionReady);
```

Comment:
I think I was wrong when I said we have single point of return :D We should also free these when no exception is thrown. That is, at the end of the `PG_TRY()` block before `PG_CATCH()`. I think we should probably refactor this big function as it is not very easy to read.

Reply:
oops, for some reason I thought this was a `try-finally`.

Reply:
Completely agree that this function could use a refactor. It would probably be easier to understand if it used a hashtable of connections instead of an array.